src/gwt/src/org/rstudio/studio/client/application/ui/addins/AddinsToolbarButton.java

```diff
@@ -0,0 +1,278 @@
+import org.rstudio.studio.client.workbench.addins.Addins.RAddins;
+import org.rstudio.studio.client.workbench.addins.AddinsCommandManager;
+import org.rstudio.studio.client.workbench.commands.Commands;
+
+import com.google.gwt.core.client.Scheduler;
+import com.google.gwt.core.client.Scheduler.ScheduledCommand;
+import com.google.gwt.dom.client.Element;
+import com.google.gwt.dom.client.Style;
+import com.google.gwt.dom.client.Style.Unit;
+import com.google.gwt.event.dom.client.ClickEvent;
+import com.google.gwt.event.dom.client.ClickHandler;
+import com.google.gwt.event.logical.shared.AttachEvent;
+import com.google.gwt.user.client.DOM;
+import com.google.gwt.user.client.Event;
+import com.google.gwt.user.client.EventListener;
+import com.google.gwt.user.client.Timer;
+import com.google.gwt.user.client.ui.Label;
+import com.google.gwt.user.client.ui.MenuItem;
+import com.google.inject.Inject;
+
+public class AddinsToolbarButton extends ToolbarButton
+{
+   public AddinsToolbarButton()
+   {
+      super("Addins",
+            CoreResources.INSTANCE.iconEmpty(),
+            new ScrollableToolbarPopupMenu(),
+            false);
+      
+      RStudioGinjector.INSTANCE.injectMembers(this);
+      
+      menu_ = getMenu();
+      
+      menu_.setAutoHideRedundantSeparators(false);
+      
+      addClickHandler(new ClickHandler()
+      {
+         @Override
+         public void onClick(ClickEvent event)
+         {
+            populate();
+         }
+      });
+      
+      addAttachHandler(new AttachEvent.Handler()
+      {
+         @Override
+         public void onAttachOrDetach(AttachEvent event)
+         {
+            if (event.isAttached())
+            {
+               menuWidth_ = menu_.getMenuTableElement().getOffsetWidth();
+            }
+            else
+            {
+               searchWidget_.setValue("");
+            }
+         }
+      });
+      
+      searchWidget_ = new SearchWidget();
+      searchValueChangeTimer_ = new Timer()
+      {
+         @Override
+         public void run()
+         {
+            onSearchValueChange();
+         }
+      };
+      
+      searchEl_ = searchWidget_.getElement();
+      
+      searchEl_.getStyle().setMarginLeft(10, Unit.PX);
+      searchEl_.getStyle().setMarginRight(10, Unit.PX);
+      searchEl_.getStyle().setMarginTop(-2, Unit.PX);
+      
+      DOM.sinkEvents(searchEl_, Event.KEYEVENTS);
+      DOM.setEventListener(searchEl_, new EventListener()
+      {
+         @Override
+         public void onBrowserEvent(Event event)
+         {
+            searchValueChangeTimer_.schedule(200);
+         }
+      });
+      
+   }
+   
+   private void populate()
+   {
+      menu_.clearItems();
+      
+      Map<String, List<RAddin>> addinsByPkg = organizeAddins();
+      
+      if (addinsByPkg.isEmpty())
+      {
+         populateEmptyMenu();
+         return;
+      }
+      
+      MapUtil.forEach(addinsByPkg, new MapUtil.ForEachCommand<String, List<RAddin>>()
```

Comment:
The addins are rendered in a scrollable menu clamped at 300px in height (https://github.com/rstudio/rstudio/blob/master/src/gwt/src/org/rstudio/core/client/widget/ScrollableToolbarPopupMenu.java#L76-L79), which should accommodate this.